targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/api/lp_ticker_api.c

```diff
@@ -0,0 +1,269 @@
+{
+    /* process RTC interrupts (cleared by driver) */
+    if (ADI_RTC_ALARM_INT & nEvent) {
+        lp_ticker_irq_handler();
+    }
+}
+
+
+/** Get low power ticker's data
+ *
+ * @return The low power ticker data
+ */
+/*
+const ticker_data_t* get_lp_ticker_data()
+{
+}*/
+
+
+/* HAL lp ticker */
+
+/** Initialize the low power ticker
+ *
+ */
+void lp_ticker_init()
+{
+    // open the rtc device
+    adi_rtc_Open(1, rtc1_memory, ADI_RTC_MEMORY_SIZE, &hRTC1_Device);
+
+    // set the prescaler value
+    adi_rtc_SetPreScale(hRTC1_Device, RTC_PRESCALER);
+
+    // register call back for events
+    adi_rtc_RegisterCallback(hRTC1_Device, rtc1_Callback, hRTC1_Device);
+
+    // reset the RTC counter
+    adi_rtc_SetCount(hRTC1_Device, 0);
+
+    // set the trim value
+    adi_rtc_SetTrim(hRTC1_Device, ADI_RTC_TRIM_INTERVAL_14, ADI_RTC_TRIM_1, ADI_RTC_TRIM_SUB);
+
+    adi_rtc_EnableTrim(hRTC1_Device, true);
+
+    // enable the RTC
+    adi_rtc_Enable(hRTC1_Device, true);
+}
+
+/** Read the current counter
+ *
+ * @return The current timer's counter value in microseconds
+ */
+uint32_t lp_ticker_read()
+{
+    uint32_t count;
+    float t;
+
+    // get current count
+    adi_rtc_GetCount(hRTC1_Device, &count);
+
+    // Throw away top 5 bits to avoid overflow and convert ticks to us
+    count &= MAX_TICK_MASK;
+    t = (float)count * TIME_US_PER_TICK;
+
+    return (uint32_t)t;
+}
+
+/** Set interrupt for specified timestamp
+ *
+ * @param timestamp The time in microseconds to be set
+ */
+void lp_ticker_set_interrupt(timestamp_t timestamp)
+{
+    uint32_t rtcCount, trunc_rtcCount, set_tick, tick_delta, alarm_tick;
+
+    // compute the tick value based on the given alarm time
+    set_tick = (uint32_t)((float)(timestamp) / TIME_US_PER_TICK);
+
+    // get current count
+    adi_rtc_GetCount(hRTC1_Device, &rtcCount);
+
+    // compute the number of ticks required for the alarm
+    trunc_rtcCount = rtcCount & MAX_TICK_MASK;
+
+    // if set_tick is less than the current RTC count, the counter has
+    // missed the desired time, so set tick_delta to 0
+    if (set_tick >= trunc_rtcCount) {
+        tick_delta = set_tick - trunc_rtcCount;
+    } else {
+        tick_delta = 0;
+    }
+
+    // compute the absolute RTC alarm count required
+    alarm_tick = rtcCount + tick_delta;
+
+    // check if the desired alarm duration is less than TICKS_TO_ENABLE_ALARM,
+    // if so just wait it out rather than setting the alarm
+    if (tick_delta <= TICKS_TO_ENABLE_ALARM) {
+        // otherwise if the alarm time is less than the current RTC count + the time
+        // it takes to enable the alarm, just wait until the desired number of counts
+        // has expired rather than using the interrupt, then call the user ISR directly.
+        do {
+            adi_rtc_GetCount(hRTC1_Device, &rtcCount);
+        } while (rtcCount < alarm_tick);
+
+        rtc1_Callback(NULL, ADI_RTC_ALARM_INT, NULL);
```

Comment:
is this calling the callback directly? If yes, this can lead  to stackoverflow. IT is recommended to fire an interrupt instead (set pending bit for instance - as fire_interrupt is doing).

Reply:
Yes it is calling the callback directly here because we are not setting up an alarm interrupt due to the long latency required for setting up this interrupt.

Reply:
Use Nvic Set Pending flag. Otherwise as noted above, we would end up with stackoverflow

Reply:
OK. I think it will take me sometime to review and make this change. I don't want to hold up the merge, is it possible to open an issue somewhere so I can submit the changes later?

Reply:
I am still not certain if I read the code correctly. From what I can tell, set interrupt is always being set with the time in the future (the upper layer can distinguish what is in the past as using 64bit counter value), thus there is fire interrupt in that case.

Reply:
> Yes it is calling the callback directly here because we are not setting up an alarm interrupt due to the long latency required for setting up this interrupt.
  
@c1728p9 Please review this code block

Reply:
Does this take 10 (TICKS_TO_ENABLE_ALARM) 32KHz ticks to set the lpticker interrupt? If so this timer is not suitable for a low power ticker and will fail testing.

Reply:
No, it takes about 4 32khz ticks (plus some other small overheads) to enable the alarm, and the adi_rtc_GetCount takes 1 32khz tick, so total of 5 ticks (around 150-200us). The issue with the lp ticker test is that even before the ticker alarm is set for the alarm value under test (i.e. 500us), the test has already called lp_ticker_read (which invokes adi_rtc_GetCount) multiple times which all take up time (not sure why the test needs to do that), so by the time it gets around to set the alarm the 500us test time is almost expired, hence we need the wait loop instead of setting the alarm. For alarms with 1ms or longer, there is no such issue. I guess the question comes down to whether a 1ms resolution for the lp ticker is sufficient. Last time I was told 1ms resolution would be adequate. Please let me know if this is not the case.

Reply:
@dave-wu Please if you find any issues with the tests (as you pinpointed above), can you capture it as new issue (with details for us to reproduce and understand the issue), would be appreciated

Reply:
This doesn't meet timing requirements for tickers and will fail HAL Specification testing because of it:
https://github.com/ARMmbed/mbed-os/blob/3c0fb931bd9b968c088a2a96e670eba55a0935bc/hal/us_ticker_api.h#L74-L75

The low power ticker should be turned off since it is not suitable for mbed-os's use case.

Reply:
Mmm, I see. I will remove this functionality for now.

Reply:
@c1728p9 I noticed that the comment regarding the ticker functions should take less than 20us to complete is for us ticker. Is the requirement the same for the lp ticker? If so it looks like a quite difficult criteria to meet for the lp ticker as the RTC is usually in a much slower clock domain and it takes quite long to synchronize between clock domains? I would think the purpose of the lp ticker is for alarms that are relatively long, otherwise you'd just use the us ticker?

Reply:
Hi @dave-wu this applies to lp tickers as well. The LP ticker is reschedule from an interrupt handler and with interrupts disabled, so with this kind of synchronization delay it can cause major system instabilities. CC @pan-